CodeU_Assignement_5/src/codeu_assignement_5/KemUnknownLanguage.java

```diff
@@ -0,0 +1,145 @@
+                    dict.add(c);
+                }
+            }
+        }
+        
+        return dict;
+    }
+    
+    
+    /*
+     * Inputs: All words with the same prefix (all the first prefixSize chars are the same!)
+     * Output: Per common prefix, gets the order of the chars
+     *
+     * Note: Assume the input is valid here!
+    */
+    private List<Character> getBasicAlphaBeit(String[] words, int prefixSize) {
+        List<Character> ret = new ArrayList(); 
+        if (words == null) {
+            return ret;
+        }
+        
+        for (String w : words) {
+            if ((w.length() > prefixSize) && (!ret.contains(w.charAt(prefixSize)))) {
+                ret.add(w.charAt(prefixSize));
+            }
+        }   
+        
+        return ret;
+    }
+    
+    /*
+     * Gives the prefix of a char that isn't in the dictionary
+    */
+    private int getSizeOfPrefix(String w, List<Character> dict) {
+        for (int i=0; i < w.length(); i++) {
+            if (!dict.contains(w.charAt(i))) {
+                return i;
+            }
+        }
+        
+        return w.length();
+    }
+    
+    /*
+     * Input: the original set of words, current word we are working on, its prefix size
+     * Output: the set of the words with this prefix
+    */
+    private String[] getWordsWithSamePrefix(String[] alienWords, int curr, int prefixSize) {
+        ArrayList<String> tempDict = new ArrayList();
+        tempDict.add(alienWords[curr]);
+        
+        String prefix = alienWords[curr].substring(0, prefixSize);
+        
+        /* Forward search */
+        for (int i=curr+1; i<alienWords.length; i++) {
+            if (!alienWords[i].startsWith(prefix)) {
+                break;
+            }
+            
+            tempDict.add(alienWords[i]);
+        }
+        
+        /* Backward search */
+        for (int i=curr-1; i >= 0; i--) {
+            if (!alienWords[i].startsWith(prefix)) {
+                break;
+            }
+            
+            tempDict.add(0,alienWords[i]); 
+        }
+        
+        return tempDict.toArray(new String[tempDict.size()]);
+    }
+    
+    /*
+     * Get two dictionary with common latters and merge them into the first dict,
+     * if none, cannot know, add all these cases later on
+    */
+    private void mergeDict(List<Character> dict1, List<Character> dict2) {
+        int from2 = 0;  
+        for (int i=0; i < dict2.size(); i++) {
+            if (dict1.contains(dict2.get(i))) {
+                /* Find common point, merge here! */
+                dict1.addAll(dict1.indexOf(dict2.get(i)), dict2.subList(from2, i));
+                from2 = i+1;
+            }        
+        }
+    }
+    
+    private boolean addToDictByPrefix(String[] alienWords, List<Character> dict) {
+        boolean hasReachMaxPrefix = true;
+        for (int i=0; i < alienWords.length; i++) {
+            int prefixSize = getSizeOfPrefix(alienWords[i], dict);
```

Comment:
The code changed quite a lot, I tried as possible to make it more clear and more general (you had some comment earlier regarding this). The idea here was to only pass on letters which we don't know their location.

It wasn't such a good idea, since anyhow we need to go over all the combination and then we need an outer loop, in addition it can be buggy in case we need to move the letter to be before, if you already push it to the end (if I remember right). It is a bit too complicated. I agree :-)

In the new version I removed it completely and remove the outer while loop...